src/fontra/views/fontinfo/panel-sources.js

```diff
@@ -1,14 +1,604 @@
+    const root = { sources: this.sources };
+    const changes = recordChanges(root, (root) => {
+      delete root.sources[this.sourceIdentifier];
+    });
+    if (changes.hasChange) {
+      this.postChange(changes.change, changes.rollbackChange, undoLabel);
+      this.setupUI();
+    }
+  }
+
+  toggleShowHide() {
+    const el = this.querySelector("#open-close-icon");
+    el.style.transform =
+      el.style.transform === "rotate(180deg)" ? "rotate(0deg)" : "rotate(180deg)";
+
+    this.style.maxHeight = this.style.maxHeight === "100%" ? "70px" : "100%";
+    for (const child of this.children) {
+      if (!child.style.maxHeight) {
+        child.style.maxHeight = "100%";
+      } else {
+        delete child.style;
+      }
+    }
+  }
+
+  _updateContents() {
+    const models = this.models;
+
+    for (const key in models) {
+      this.controllers[key] = new ObservableController(models[key]);
+      this.controllers[key].addListener((event) => {
+        if (key == "location") {
+          if (!this.checkSourceLocation(event.key, event.newValue)) {
+            return;
+          }
+        }
+
+        if (event.key == "name") {
+          if (!this.checkSourceEntry("name", undefined, event.newValue.trim())) {
+            return;
+          }
+        }
+
+        this.editSource((source) => {
+          if (key == "general") {
+            if (typeof event.key == "string") {
+              source[event.key] = event.newValue.trim();
+            } else {
+              source[event.key] = event.newValue;
+            }
+          } else {
+            if (key == "verticalMetrics") {
+              if (event.key.startsWith("value-")) {
+                source[key][event.key.slice(6)].value = event.newValue;
+              } else {
+                source[key][event.key.slice(5)].zone = event.newValue;
+              }
+            } else {
+              source[key][event.key] = event.newValue;
+            }
+          }
+        }, `edit source ${key} ${event.key}`);
+      });
+    }
+
+    this.innerHTML = "";
+    this.append(
+      html.createDomElement("icon-button", {
+        class: "fontra-ui-font-info-icon open-close-icon",
+        style: "translate: 120ms; transform: rotate(180deg)",
+        id: "open-close-icon",
+        src: "/tabler-icons/chevron-up.svg",
+        onclick: (event) => this.toggleShowHide(),
+      })
+    );
+
+    for (const key in models) {
+      this.append(html.div({ class: "fontra-ui-font-info-header" }, [translate(key)]));
+    }
+
+    this.append(
+      html.createDomElement("icon-button", {
+        "class": "fontra-ui-font-info-icon",
+        "src": "/tabler-icons/trash.svg",
+        "onclick": (event) => this.deleteSource(),
+        "data-tooltip": translate("sources.delete-source"),
+        "data-tooltipposition": "left",
+      })
     );
+
+    this.append(html.div()); // empty cell for grid with arrow
+
+    for (const key in models) {
```

Comment:
Likewise, this whole block can be much better written like this:
```js
    this.append(buildElement(this.controllers.general));
    this.append(buildElementLocations(this.controllers.location, this.fontAxes));
    this.append(buildElementVerticalMetrics(this.controllers.verticalMetrics));
```

Reply:
done.